src/execution/operator/join/physical_piecewise_merge_join.cpp

```diff
@@ -190,7 +344,174 @@ class PiecewiseMergeJoinState : public OperatorState {
+	scan.entry_idx = entry_idx;
+	return scan.RadixPtr();
+}
+
+static idx_t MergeJoinSimpleBlocks(PiecewiseMergeJoinState &lstate, MergeJoinGlobalState &rstate, bool *found_match,
+                                   const ExpressionType comparison) {
+	const auto cmp = MergeJoinComparisonValue(comparison);
+
+	// The sort parameters should all be the same
+	auto &lsort = *lstate.lhs_global_state;
+	auto &rsort = rstate.rhs_global_sort_state;
+	D_ASSERT(lsort.sort_layout.all_constant == rsort.sort_layout.all_constant);
+	const auto all_constant = lsort.sort_layout.all_constant;
+	D_ASSERT(lsort.external == rsort.external);
+	const auto external = lsort.external;
+
+	// There should only be one sorted block if they have been sorted
+	D_ASSERT(lsort.sorted_blocks.size() == 1);
+	SBScanState lread(lsort.buffer_manager, lsort);
+	lread.sb = lsort.sorted_blocks[0].get();
+
+	const idx_t l_block_idx = 0;
+	idx_t l_entry_idx = 0;
+	const auto lhs_not_null = lstate.lhs_count - lstate.lhs_has_null;
+	MergeJoinPinSortingBlock(lread, l_block_idx);
+	auto l_ptr = MergeJoinRadixPtr(lread, l_entry_idx);
+
+	D_ASSERT(rsort.sorted_blocks.size() == 1);
+	SBScanState rread(rsort.buffer_manager, rsort);
+	rread.sb = rsort.sorted_blocks[0].get();
+
+	const auto cmp_size = lsort.sort_layout.comparison_size;
+	const auto entry_size = lsort.sort_layout.entry_size;
+
+	idx_t right_base = 0;
+	for (idx_t r_block_idx = 0; r_block_idx < rread.sb->radix_sorting_data.size(); r_block_idx++) {
+		// we only care about the BIGGEST value in each of the RHS data blocks
+		// because we want to figure out if the LHS values are less than [or equal] to ANY value
+		// get the biggest value from the RHS chunk
+		MergeJoinPinSortingBlock(rread, r_block_idx);
+
+		auto &rblock = rread.sb->radix_sorting_data[r_block_idx];
+		const auto r_not_null = SortedBlockNotNull(right_base, rblock.count, rstate.rhs_count - rstate.rhs_has_null);
+		if (0 == r_not_null) {
+			break;
+		}
+		const auto r_entry_idx = r_not_null - 1;
+		right_base += rblock.count;
+
+		auto r_ptr = MergeJoinRadixPtr(rread, r_entry_idx);
+
+		// now we start from the current lpos value and check if we found a new value that is [<= OR <] the max RHS
+		// value
+		while (true) {
+			int comp_res;
+			if (all_constant) {
+				comp_res = memcmp(l_ptr, r_ptr, cmp_size);
+			} else {
+				comp_res = Comparators::CompareTuple(lread, rread, l_ptr, r_ptr, lsort.sort_layout, external);
+			}
+
+			if (comp_res <= cmp) {
+				// found a match for lpos, set it in the found_match vector
+				found_match[l_entry_idx] = true;
+				l_entry_idx++;
+				l_ptr += entry_size;
+				if (l_entry_idx >= lhs_not_null) {
+					// early out: we exhausted the entire LHS and they all match
+					return 0;
+				}
+			} else {
+				// we found no match: any subsequent value from the LHS we scan now will be bigger and thus also not
+				// match move to the next RHS chunk
+				break;
+			}
+		}
```

Comment:
We can remove most of the branching in this while loop (albeit with some duplicate code), and we should use `FastMemcmp` because the size of what's being compared stays constant throughout the loop, e.g.:
```c++
if (all_constant) {
	int comp_res = FastMemcmp(l_ptr, r_ptr, cmp_size);
	while (comp_res <= cmp) {
		// found a match for lpos, set it in the found_match vector
		found_match[l_entry_idx] = true;
		l_entry_idx++;
		if (l_entry_idx >= lhs_not_null) {
			// early out: we exhausted the entire LHS and they all match
			return 0;
		}
		l_ptr += entry_size;
		comp_res = FastMemcmp(l_ptr, r_ptr, cmp_size);
	}
} else {
	int comp_res = Comparators::CompareTuple(lread, rread, l_ptr, r_ptr, lsort.sort_layout, external);
	while (comp_res <= cmp) {
		// found a match for lpos, set it in the found_match vector
		found_match[l_entry_idx] = true;
		l_entry_idx++;
		if (l_entry_idx >= lhs_not_null) {
			// early out: we exhausted the entire LHS and they all match
			return 0;
		}
		l_ptr += entry_size;
		comp_res = Comparators::CompareTuple(lread, rread, l_ptr, r_ptr, lsort.sort_layout, external);
	}
}
```
I would expect that this would speed up the loop, not sure how much.

I would also be happy to profile/tweak this code myself if you'd prefer that!

Reply:
Thanks for the comments!

We have one benchmark (`benchmark/micro/join/range_join.benchmark`) that I was using for the comparison above. IIRC the majority (~50%) of the time was in `memcmp` and another 25% was in the method containing it.  

The comment about `FastMemcmp` is interesting and I can benchmark that on the same machine later. But if you want to do your own comparisons, please do! And does it really matter that the comparison is in the loop - won't branch prediction handle it?

Aside from using the new sorting code, my other reason for this rewrite is to come up with a fair comparison for `IEJoin` and prototype the techniques it would use. As part of that I will be adding support for multiple predicates, which would basically just be evaluating the additional expressions  as a final filtering step.

Reply:
I've played around with this for a bit, and you can probably ignore my suggestion about eliminating branches, this does not seem to improve performance. The main thing is using `FastMemcmp` instead of `memcmp`, this speeds up the benchmark quite a bit.